### ISSUE TITLE
automation: test failed

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#878]
Building Eclipse SmartHome using maven failed for me (at least once).

It could be related to #820 and https://bugs.eclipse.org/bugs/show_bug.cgi?id=479215
But perhaps it is also an error in the automation implementation.

Could someone of the automation guys have a look at?

```text
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.eclipse.smarthome.automation.event.RuleEventTest
11:42:34.600 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'com.sun.proxy.$Proxy7' has been added.
11:42:34.622 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.core.items.ManagedItemProvider' has been added.
11:42:34.634 [main] DEBUG o.e.s.a.i.c.p.AutomationResourceBundlesEventQueue - Process bundle event 1, for automation bundle 'org.eclipse.smarthome.automation.module.core' 
11:42:34.649 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'org.eclipse.smarthome.automation.core.internal.ManagedRuleProvider' has been added.
11:42:34.684 [Automation Provider Processing Queue] DEBUG o.e.s.a.i.c.p.RuleResourceBundleImporter - Parse rules from bundle 'org.eclipse.smarthome.automation.module.core' 
11:42:34.742 [main] INFO  o.e.s.automation.event.RuleEventTest - Rule created: myRule21
11:42:34.760 [main] DEBUG o.e.s.a.core.internal.RuleEngine - Added rule 'myRule21'
11:42:34.763 [EventAdmin Async Event Dispatcher Thread] DEBUG o.e.s.core.common.ThreadPoolManager - Created thread pool 'safeCall' with size 5-10
11:42:34.763 [main] DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory - create composite module:org.eclipse.smarthome.automation.core.internal.RuntimeCondition@75c9e76b, of rule: myRule21
11:42:34.764 [main] DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory - create composite module:org.eclipse.smarthome.automation.Condition@7446d8d5, of rule: myRule21
11:42:34.765 [main] DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory - Set handler to composite module:org.eclipse.smarthome.automation.Condition@7446d8d5 -> org.eclipse.smarthome.automation.core.internal.composite.CompositeConditionHandler@47db5fa5
11:42:34.765 [main] DEBUG o.e.s.a.c.i.c.CompositeModuleHandlerFactory - Set handler to composite module:org.eclipse.smarthome.automation.core.internal.RuntimeCondition@75c9e76b -> org.eclipse.smarthome.automation.core.internal.composite.CompositeConditionHandler@354fc8f0
11:42:34.771 [main] DEBUG o.e.s.c.c.r.AbstractManagedProvider - Added new element myRule21 to ManagedRuleProvider.
11:42:34.778 [ESH-safeCall-1] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/state
11:42:34.788 [main] INFO  o.e.s.a.core.internal.RuleEngine - The rule rId = myRule21 is already enabled
11:42:34.797 [ESH-safeCall-2] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/state
11:42:34.800 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/added
11:42:34.856 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - Event: smarthome/items/myMotionItem2/state
11:42:34.858 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/state
11:42:34.860 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemStateConditionHandler - ItemStateCondition 'ItemStateCondition3'checking if myPresenceItem2 (State=ON) = ON
11:42:34.861 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemStateConditionHandler - ConditionSatisfied --> true
11:42:34.861 [pool-2-thread-1] DEBUG o.e.s.a.m.c.h.ItemPostCommandActionHandler - Executing ItemPostCommandAction on Item myLampItem2 with Command ON
11:42:34.864 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - Event: smarthome/items/myLampItem2/state
11:42:34.866 [pool-2-thread-1] DEBUG o.e.s.a.core.internal.RuleEngine - The rule 'myRule21' is executed.
11:42:34.868 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/state
11:42:34.971 [main] INFO  o.e.s.automation.event.RuleEventTest - myLampItem2 State: ON
11:42:34.981 [main] DEBUG o.e.s.c.c.r.AbstractManagedProvider - Removed element myRule21 from ManagedRuleProvider.
11:42:34.985 [ESH-safeCall-3] INFO  o.e.s.automation.event.RuleEventTest - RuleEvent: smarthome/rules/myRule21/removed
11:42:34.994 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider '$Proxy7' has been removed.
11:42:34.994 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'ManagedItemProvider' has been removed.
11:42:34.997 [main] DEBUG o.e.s.c.c.registry.AbstractRegistry - Provider 'ManagedRuleProvider' has been removed.
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.899 sec <<< FAILURE! - in org.eclipse.smarthome.automation.event.RuleEventTest
testRuleEvents(org.eclipse.smarthome.automation.event.RuleEventTest)  Time elapsed: 0.533 sec  <<< ERROR!
java.util.ConcurrentModificationException: null
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:901)
	at java.util.ArrayList$Itr.next(ArrayList.java:851)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.find(DefaultGroovyMethods.java:3059)
	at org.codehaus.groovy.runtime.dgm$175.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:271)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:53)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
	at org.eclipse.smarthome.automation.event.RuleEventTest$_testRuleEvents_closure17.doCall(RuleEventTest.groovy:175)
	at org.eclipse.smarthome.automation.event.RuleEventTest$_testRuleEvents_closure17.doCall(RuleEventTest.groovy)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:324)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:278)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:54)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy:174)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:207)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:133)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:141)
	at org.eclipse.smarthome.automation.event.RuleEventTest.testRuleEvents(RuleEventTest.groovy:174)


Results :

Tests in error: 
  RuleEventTest.testRuleEvents:174->OSGiTest.waitForAssert:-1->OSGiTest.waitForAssert:174 » ConcurrentModification

Tests run: 1, Failures: 0, Errors: 1, Skipped: 0
```

**Comment by marinmitev** [comment]
The problem cannot be reproduced on Windows with java8 and I think it is not a problem in the test itself because it was never modified since its first version.
The source line where the problem occurs has the following code:
        waitForAssert({
            assertThat ruleEvents.find{it.topic=="smarthome/rules/myRule21/removed"}, is(notNullValue())
        })

**Comment by maggu2810** [comment]
I looked at the source code now, and I am pretty sure it is a bug in the test.
I remember to fixed already a similar one.

You are using the ruleEvents list and look for a member. This is done by using an iterator.
The list is modified by the event handler running in a separate thread.
So this list is iterated the same time as it can be modified and this is a common Concurrent Modification candidate.

**Comment by marinmitev** [comment]
code is fixed, thanks for the provided help

**Comment by maggu2810** [comment]
Thanks for the fix (#881).
